source/keymap.js

```diff
@@ -11,8 +11,327 @@
+						this.appID = foregroundAppInfo[i].appId;
+						break;
+					}
+				}
+				if(appID !== ''){
+					this.getAppInfo(appID);
+				}
+			}
+		},
+
+		/**
+		* @private
+		*/
+		getAppInfo: function (appID) {
+			if(scope.PalmSystem) {
+				var param = {};
+				param.id = appID;
+				this.$.getAppInfo.send(param);
+			}
+		},
+
+		/**
+		* @private
+		*/
+		getAppInfoHandler: function (inSender, inResponse) {
+			if(inResponse.appInfo !== undefined) {
+				this.enableBackHistoryAPI = !inResponse.appInfo.disableBackHistoryAPI;
+			}
+		},
+
+		/**
+		* @public
+		*/
+		getCurrentObj: function() {
+			return this._currentObj;
+		},
+
+		/**
+		* Set target object and handler of back key
+		*
+		* @param {Object} ctx - current object for back key
+		* @param {function} fn - handler of back key
+		* @private
+		*/
+		_pushBackHistory: function(ctx, fn) {
+			if (this.enableBackHistoryAPI) {
+				history.pushState({currentObjId: ctx.id}, '', '');
+			}
+			this._backHistoryStack.push({currentObj: ctx, handler: fn});
+			this._currentObj = ctx;
+			this._handler = fn;
+		},
+
+		/**
+		* Store back key request to history.
+		* If popstate is in progress, it delays until former work is done
+		*
+		* @param {Object} ctx - current object for back key
+		* @param {function} fn - handler of back key
+		* @public
+		*/
+		pushBackHistory: function(ctx, fn) {
+			if (this._isPopStateInProgress) {
+				this._pushBackQueue.push({currentObj: ctx, handler: fn});
+			} else {
+				this._pushBackHistory(ctx, fn);
+			}
+		},
+
+		/**
+		* When we press back key of R/C, window.history should be back 1 step.
+		* With synchronizing window.history and _backHistoryStack
+		* we do not want to run back key handler.
+		*
+		* @public
+		*/
+		ignorePopState: function() {
+			if (this.enableBackHistoryAPI) {
+				this._ignorePopState = true;
+				this._isPopStateInProgress = true;
+				history.go(-1);
+			}
+		},
+
+		/**
+		* @private
+		*/
+		callBackKeyHandler: function() {
+			this.isBackInProgress = true;
+			if (this._currentObj && this._handler) {
+				var fn = this._handler;
+				if (typeof fn == 'function') {
+					fn = fn.bind(this._currentObj);
+				}
+				fn();
+			}
+		},
+
+		/**
+		* @private
+		*/
+		popBackHistory: function() {
+			var bStack = this._backHistoryStack;
+			bStack.pop();
+			if (bStack.length) {
+				this._currentObj = bStack[bStack.length - 1].currentObj;
+				this._handler = bStack[bStack.length - 1].handler;
+			} else {
+				this._currentObj = this._handler = null;
+			}
+
+			this.isBackInProgress = false;
+		},
+
+		/**
+		* Dequeue _pushBackQueue
+		*
+		* @privae
+		*/
+		_dePushBackQueue: function() {
+			var queue = this._pushBackQueue,
+				length = queue.length,
+				item;
+			for (var i = 0; i < length; i++) {
+				item = queue.pop();
+				this._pushBackHistory(item.currentObj, item.handler);
+			}
+		},
+
+		/**
+		* Decide whether this popstate event should call handler or shouldn't.
+		*
```

Comment:
Hi @DavidUM, I think it would be helpful to document all of the various pop states with some details about what they represent.